TSynth/TSynth.ino

```diff
@@ -380,514 +294,143 @@ void myNoteOn(byte channel, byte note, byte velocity) {
-    case 10:
-      filterEnvelope10.noteOff();
-      ampEnvelope10.noteOff();
-      voices[9].voiceOn = false;
-      break;
-    case 11:
-      filterEnvelope11.noteOff();
-      ampEnvelope11.noteOff();
-      voices[10].voiceOn = false;
-      break;
-    case 12:
-      filterEnvelope12.noteOff();
-      ampEnvelope12.noteOff();
-      voices[11].voiceOn = false;
-      break;
-    default:
-      //Do nothing
-      break;
-  }
+// renamed from endVoice to match voiceOn
+void voiceOff(uint8_t index) {
+  if (index < 0 || index >= NO_OF_VOICES) return;
+  Oscillators[index].filterEnvelope_.noteOff();
+  Oscillators[index].ampEnvelope_.noteOff();
+  voices[index].voiceOn = false;
 }
 
 void myNoteOff(byte channel, byte note, byte velocity) {
   decNotesOn();
   if (unison == 0) {
-    endVoice(getVoiceNo(note));
+    voiceOff(getVoiceNo(note));
   } else {
-    //UNISON MODE
-    for (uint32_t i = 0; i < NO_OF_VOICES; i++) {
-      endVoice(getVoiceNo(note));
-    }
+    //UNISON MODE: disable all voices for note.
+    FOR_EACH_VOICE(voiceOff(getVoiceNo(note)))
   }
 }
 
 void allNotesOff() {
   notesOn = 0;
-  for (uint32_t v = 0; v < NO_OF_VOICES; v++) {
-    endVoice(v + 1);
-  }
+
+  FOR_EACH_VOICE(voiceOff(i))
 }
 
+// Get the index into the 'voices' and 'Oscillators' array for a given note.
+// or -1 to get the next free, or oldest active note.
+// TODO: Include a channel for multi-timbral mode.
 int getVoiceNo(int note) {
   voiceToReturn = -1;      //Initialise
   earliestTime = millis(); //Initialise to now
   if (note == -1) {
     //NoteOn() - Get the oldest free voice (recent voices may be still on release stage)
-    for (uint32_t i = 0; i < NO_OF_VOICES; i++) {
+    FOR_EACH_VOICE(
       if (!voices[i].voiceOn) {
         if (voices[i].timeOn < earliestTime) {
           earliestTime = voices[i].timeOn;
           voiceToReturn = i;
         }
       }
-    }
+    )
     if (voiceToReturn == -1) {
       //No free voices, need to steal oldest sounding voice
-      earliestTime = millis(); //Reinitialise
-      for (uint32_t i = 0; i < NO_OF_VOICES; i++) {
+      FOR_EACH_VOICE(
         if (voices[i].timeOn < earliestTime) {
           earliestTime = voices[i].timeOn;
           voiceToReturn = i;
         }
-      }
+      )
     }
-    return voiceToReturn + 1;
+    return voiceToReturn;
   } else {
     //NoteOff() - Get voice number from note
-    for (uint32_t i = 0; i < NO_OF_VOICES; i++) {
+    FOR_EACH_VOICE(
       if (voices[i].note == note && voices[i].voiceOn) {
-        return i + 1;
+        return i;
       }
-    }
+    )
     //Unison - Note on without previous note off?
-    return voiceToReturn;
+    return -1;
   }
   //Shouldn't get here, return voice 1
   return 1;
 }
 
-void updateVoice1() {
-  if (unison == 1) {
-    waveformMod1a.frequency(NOTEFREQS[voices[0].note + oscPitchA]);
-    waveformMod1b.frequency(NOTEFREQS[voices[0].note + oscPitchB] * (detune + ((1 - detune) * DETUNE[notesOn - 1][1])));

```

Comment:
`waveformMod1a` with unison == 1 is the only one which doesn't apply a detune ratio. Is that because the DETUNE array has 0.0 for DETUNE[N][0], causing this portion to be a no-op?

My `updateVoice(...)` function includes the modifier for `waveformMod1a`.

Reply:
Yes `waveformMod1a` isn't detuned.